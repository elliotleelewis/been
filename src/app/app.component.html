<mgl-map
	[attributionControl]="false"
	[minZoom]="minZoom"
	[style]="'mapbox://styles/mapbox/streets-v9'"
	[zoom]="[minZoom]"
></mgl-map>
