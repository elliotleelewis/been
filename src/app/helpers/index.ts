export * from './coordinates';
export * from './regionalizer';
