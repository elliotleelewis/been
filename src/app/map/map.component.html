<mgl-map
	logoPosition="bottom-right"
	[attributionControl]="false"
	[minZoom]="minZoom"
	[style]="'mapbox://styles/elliotleelewis/ckj1w9ro12g8h19mqhvsgcbkg'"
	[zoom]="[minZoom]"
>
	<mgl-vector-source
		id="countries"
		url="mapbox://mapbox.country-boundaries-v1"
	></mgl-vector-source>
	<mgl-layer
		id="visited-countries"
		type="fill"
		source="countries"
		sourceLayer="country_boundaries"
		[filter]="filter"
		[paint]="{ 'fill-color': '#fd7e14', 'fill-opacity': 0.7 }"
	></mgl-layer>
</mgl-map>
