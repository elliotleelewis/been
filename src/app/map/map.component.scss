:host {
	mgl-map {
		width: 100%;
		height: 100%;

		::ng-deep {
			canvas {
				outline: none;
			}
		}
	}
}
