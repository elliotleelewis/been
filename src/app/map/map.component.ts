import {
	ChangeDetectionStrategy,
	Component,
	HostBinding,
	Inject,
	type OnDestroy,
	type OnInit,
} from '@angular/core';
import { type Map } from 'mapbox-gl';
import { type Observable, distinctUntilChanged, fromEvent, map } from 'rxjs';
import { SubSink } from 'subsink';

import {
	type Coordinate,
	type NestedCoordinates,
	getCoordsCenter,
} from '../helpers';
import { CountriesService } from '../services/countries.service';

@Component({
	selector: 'app-map',
	templateUrl: './map.component.html',
	styleUrls: ['./map.component.scss'],
	changeDetection: ChangeDetectionStrategy.OnPush,
})
export class MapComponent implements OnInit, OnDestroy {
	@HostBinding('class')
	class = 'flex flex-col';

	readonly darkThemeUrl = 'mapbox://styles/mapbox/dark-v11';
	readonly lightThemeUrl = 'mapbox://styles/mapbox/light-v11';
	readonly prefersDark = window.matchMedia('(prefers-color-scheme: dark)')
		.matches;

	map?: Map;
	center?: Coordinate;
	minZoom = 1.8;

	private _subs = new SubSink();

	constructor(
		@Inject(CountriesService) private countriesService: CountriesService,
	) {}

	static get prefersDark$(): Observable<boolean> {
		return fromEvent<MediaQueryListEvent>(
			window.matchMedia('(prefers-color-scheme: dark)'),
			'change',
		).pipe(
			map((event) => event.matches),
			distinctUntilChanged(),
		);
	}

	get countriesSelected$(): Observable<string[]> {
		return this.countriesService.countries$.pipe(
			map((countries) =>
				countries.filter((c) => c.selected).map((c) => c.iso3166),
			),
		);
	}

	ngOnInit(): void {
		this._subs.sink = this.countriesService.focus$.subscribe((country) => {
			const query = this.map?.querySourceFeatures('countries', {
				sourceLayer: 'country_boundaries',
				filter: ['==', ['get', 'iso_3166_1'], country],
			});
			const feature = query?.[0];
			if (feature) {
				const { coordinates } = feature.geometry as {
					coordinates?: NestedCoordinates;
				};
				if (coordinates) {
					this.center = getCoordsCenter(coordinates);
				}
			}
		});

		this._subs.sink = MapComponent.prefersDark$.subscribe((prefersDark) => {
			this.map?.setStyle(
				prefersDark ? this.darkThemeUrl : this.lightThemeUrl,
			);
		});
	}

	ngOnDestroy(): void {
		this._subs.unsubscribe();
	}

	onMapLoad(map: Map): void {
		this.map = map;
	}
}
