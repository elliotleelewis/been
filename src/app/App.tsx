import { type FC } from 'react';

import { CountriesProvider } from './contexts/countries-context';
import { Header } from './header/header';
import { Map } from './map/map';
import { Menu } from './menu/menu';

export const App: FC = () => {
	return (
		<CountriesProvider>
			<div className="flex h-full flex-col-reverse sm:flex-row">
				<div className="flex basis-1/3 flex-col overflow-auto dark:bg-zinc-900 dark:text-white">
					<Menu header={<Header show="tablet" />} />
				</div>
				<div className="basis-2/3">
					<Map header={<Header />} />
				</div>
			</div>
		</CountriesProvider>
	);
};
