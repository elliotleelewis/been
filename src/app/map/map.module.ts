import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { MAPBOX_API_KEY, NgxMapboxGLModule } from 'ngx-mapbox-gl';

import { ENVIRONMENT } from '../../environments/environment';

import { MapComponent } from './map.component';

@NgModule({
	declarations: [MapComponent],
	imports: [BrowserModule, NgxMapboxGLModule],
	exports: [MapComponent],
	providers: [
		{
			provide: MAPBOX_API_KEY,
			useValue: ENVIRONMENT.apiKeyMapbox,
		},
	],
})
export class MapModule {}
