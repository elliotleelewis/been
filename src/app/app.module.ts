import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { BrowserModule } from '@angular/platform-browser';
import { MAPBOX_API_KEY, NgxMapboxGLModule } from 'ngx-mapbox-gl';

import { ENVIRONMENT } from '../environments/environment';

import { AppComponent } from './app.component';
import { HeaderComponent } from './header/header.component';
import { MapComponent } from './map/map.component';
import { MenuComponent } from './menu/menu.component';

@NgModule({
	declarations: [AppComponent, HeaderComponent, MapComponent, MenuComponent],
	imports: [
		BrowserModule,
		FormsModule,
		ReactiveFormsModule,
		NgxMapboxGLModule,
	],
	providers: [
		{
			provide: MAPBOX_API_KEY,
			useValue: ENVIRONMENT.apiKeyMapbox,
		},
	],
	bootstrap: [AppComponent],
})
export class AppModule {}
