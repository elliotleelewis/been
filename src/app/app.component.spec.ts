import type { ComponentFixture } from '@angular/core/testing';
import { TestBed } from '@angular/core/testing';
import { MockComponents } from 'ng-mocks';

import { AppComponent } from './app.component';
import { MapComponent } from './map/map.component';
import { MenuComponent } from './menu/menu.component';

describe('AppComponent', () => {
	let component: AppComponent;
	let fixture: ComponentFixture<AppComponent>;

	beforeEach(async () => {
		await TestBed.configureTestingModule({
			declarations: [
				AppComponent,
				MockComponents(MapComponent, MenuComponent),
			],
		}).compileComponents();
	});

	beforeEach(() => {
		fixture = TestBed.createComponent(AppComponent);
		component = fixture.componentInstance;
		fixture.detectChanges();
	});

	it('should create', () => {
		expect(component).toBeTruthy();
	});

	it('should render', () => {
		expect(fixture).toMatchSnapshot();
	});
});
