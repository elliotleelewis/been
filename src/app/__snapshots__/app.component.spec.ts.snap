// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent should render 1`] = `
<app-root>
  <div
    class="flex h-full"
  >
    <app-menu
      class="basis-1/3"
    />
    <app-map
      class="basis-2/3"
    />
  </div>
</app-root>
`;
